src/utils/jsonlogic.ts

```diff
@@ -0,0 +1,169 @@
+  );
+  // formVariables contain user defined and component variables
+  // but dataType `array` and `object` are incomplete
+  // overwrite the component variables
+  Object.entries(components).map(
+    ([key, component]) => (data[key] = dataTypeForComponent(component))
+  );
+
+  return (logic, expected = undefined) => {
+    // We don't evaluate logic, we just look at types.
+    // "===" forces the logic expression align with the expectancy if defined
+    const result = infer(expected !== undefined ? {'===': [logic, expected]} : logic, data);
+    return result.startsWith('result type:') ? '' : result;
+  };
+};
+
+// TODO use infer if serviceFetchConfiguration.data_mapping_type == 'JsonLogic'
+const dataTypeForVariableDefinition = ({initialValue, dataType}: VariableDefinition): JSONValue =>
+  initialValue !== null
+    ? initialValue // may be something more complete than [] or {} ... TODO prefillAttribute too?
+    : {
+        string: '',
+        int: 1,
+        float: 1.1,
+        array: [],
+        object: {},
+        date: '',
+        datetime: '',
+        time: '',
+        boolean: true,
+      }[dataType];
+
+const dataTypeForComponent = (
+  component:
+    | AnyComponentSchema
+    | {
+        type: 'map' | 'editgrid' | 'password' | 'signature'; // Not yet implemented in this builder
+        multiple?: boolean;
+        defaultValue: JSONValue;
+      }
+): JSONValue => {
+  // For now return example values as accepted by InferNoLogic
+  // But example values cannot distinguish arrays from tuples!
+  const value = {
+    address: {
+      postcode: '',
+      houseNumber: '',
+      houseLetter: '',
+      houseNumberAddition: '',
+    },
+    currency: 1,
+    number: 1,
+    checkbox: true,
+    //@ts-ignore selectboxes always have component.defaultValue (this does work when rewritten as a lengthy switch/case)
+    selectboxes: component.defaultValue,
+    npFamilyMembers: {}, // TODO record type
+    map: [1, 1], // TODO tuple type
+    bsn: 'string',
+    columns: null, // layout component
+    content: null, // layout component
+    date: 'date', // TODO string for now
+    datetime: 'datetime', // TODO string for now
+    editgrid: [{}], // TODO inspect the component
+    fieldset: null, // layout component
+    email: 'string',
+    file: [
+      {
+        data: {baseUrl: '', form: '', name: '', project: '', size: 0, url: ''},
+        name: '',
+        originalName: '',
+        size: 0,
+        storage: '',
+        type: '', // mime
+        url: '',
+      },
+    ],
+    iban: 'string',
+    licenseplate: 'string',
+    password: 'string',
+    phoneNumber: 'string',
+    postcode: 'string',
+    radio: 'string',
+    select: 'string',
+    signature: 'string', // data-url
+    textarea: 'string',
+    textfield: 'string',
+    time: 'time', // TODO string for now
+  }[component.type];
+  return 'multiple' in component && component.multiple ? [value] : value;
```

Comment:
file multiple is still just an array of objects, I think this produces `[[{...}]]` now?

Also a more correct check is `hasOwnProperty(component, 'multiple')` which also adds the type marker that `component.multiple` actually exists. It prevents some accidents in the prototype chain where `multiple` could be defined as a method for example, which has a different meaning.

Reply:
Yeah. using `hasOwnProperty` doesn't type check;